File: multithreading/cppnuts/v1.py
import threading
import time

OddSum = 0
EvenSum = 0

start = 0
end = 190_000_000

def find_even(start, end):
    global EvenSum
    for i in range(start, end + 1):
        if i % 2 == 0:
            EvenSum += i

def find_odd(start, end):
    global OddSum
    for i in range(start, end + 1):
        if i % 2 == 1:
            OddSum += i

# --- Threaded Execution ---
start_time = time.time()

t1 = threading.Thread(target=find_even, args=(start, end))
t2 = threading.Thread(target=find_odd, args=(start, end))

t1.start()
t2.start()

t1.join()
t2.join()

threaded_time = time.time() - start_time

print(f"Threaded OddSum: {OddSum}")
print(f"Threaded EvenSum: {EvenSum}")
print(f"Threaded Time: {threaded_time:.2f} seconds")

# --- Reset and Non-threaded Execution ---
OddSum = 0
EvenSum = 0

start_time = time.time()
find_even(start, end)
find_odd(start, end)
non_threaded_time = time.time() - start_time

print(f"\nNon-threaded OddSum: {OddSum}")
print(f"Non-threaded EvenSum: {EvenSum}")
print(f"Non-threaded Time: {non_threaded_time:.2f} seconds")

File: multithreading/cppnuts/v1.cpp
#include<iostream>
#include <ratio>
#include<thread>
#include<chrono>
#include<algorithm>

using namespace std;
using namespace std::chrono;

typedef unsigned long long ull;

ull OddSum = 0;
ull EvenSum = 0;


void findEven(ull start, ull end){
    for (ull i = start; i <= end; ++i){
        if((i&1)==0){
            EvenSum += i;
        }
    }
}

void findOdd(ull start, ull end){
    for (ull i = start; i <= end; ++i){
        if((i&1)==1){
            OddSum += i;
        }
    }
}

int main(){
    ull start = 0, end=190000000;

    auto startTime = high_resolution_clock::now();
    
    std::thread t1(findEven, start, end);
    std::thread t2(findOdd, start, end);
    
    t1.join();
    t2.join();

    // findOdd( start,  end);
    // findEven(start, end);
    auto stopTime = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stopTime - startTime);

    cout << "OddSum: " << OddSum << "\n";
    cout << "EvenSum: " << EvenSum << "\n";
    cout << "Time took: " << duration.count()/1000000 << "\n";
    return 0;

}